File: ios/Classes/GeoMath.swift
import Foundation
import CoreLocation

/**
 * Centralized geospatial mathematics utility.
 * Single source of truth for geofencing algorithms across the engine.
 * Ensures algorithm parity with Kotlin (Android) implementation.
 *
 * Thread-safe: all methods are pure (no side effects).
 */
internal enum GeoMath {
    private static let EARTH_RADIUS_METERS: Double = 6371000.0

    /**
     * Check if a polygon crosses the antimeridian (±180° longitude).
     * A polygon crosses if any consecutive edge spans >180° longitude difference.
     *
     * - Parameter polygon: List of polygon vertices
     * - Returns: true if polygon crosses antimeridian, false otherwise
     */
    private static func crossesAntimeridian(_ polygon: [CLLocationCoordinate2D]) -> Bool {
        for i in 0..<polygon.count {
            let p1 = polygon[i]
            let p2 = polygon[(i + 1) % polygon.count]
            let lngDiff = abs(p2.longitude - p1.longitude)
            if lngDiff > 180.0 {
                return true
            }
        }
        return false
    }

    /**
     * Normalize longitude from [-180, 180] to [0, 360].
     * Used for antimeridian-crossing polygon handling.
     *
     * - Parameter lng: Longitude in [-180, 180]
     * - Returns: Longitude in [0, 360]
     */
    private static func normalizeLongitude(_ lng: Double) -> Double {
        return lng < 0 ? lng + 360.0 : lng
    }

    /**
     * Calculate distance between two points using Haversine formula.
     *
     * - Parameters:
     *   - point1: First coordinate
     *   - point2: Second coordinate
     * - Returns: Distance in meters
     */
    static func haversineDistance(
        point1: CLLocationCoordinate2D,
        point2: CLLocationCoordinate2D
    ) -> Double {
        return haversineDistance(
            lat1: point1.latitude,
            lng1: point1.longitude,
            lat2: point2.latitude,
            lng2: point2.longitude
        )
    }

    /**
     * Calculate distance between two points using Haversine formula.
     *
     * - Parameters:
     *   - lat1: First point latitude
     *   - lng1: First point longitude
     *   - lat2: Second point latitude
     *   - lng2: Second point longitude
     * - Returns: Distance in meters
     */
    static func haversineDistance(
        lat1: Double,
        lng1: Double,
        lat2: Double,
        lng2: Double
    ) -> Double {
        let dLat = (lat2 - lat1) * .pi / 180
        let dLng = (lng2 - lng1) * .pi / 180

        let a = sin(dLat / 2) * sin(dLat / 2) +
                cos(lat1 * .pi / 180) * cos(lat2 * .pi / 180) *
                sin(dLng / 2) * sin(dLng / 2)

        let c = 2 * atan2(sqrt(a), sqrt(1 - a))
        return EARTH_RADIUS_METERS * c
    }

    /**
     * Point-in-polygon detection using ray casting algorithm.
     * Handles antimeridian-crossing polygons by normalizing longitudes.
     *
     * - Parameters:
     *   - point: Point coordinate
     *   - polygon: List of polygon vertices
     * - Returns: true if point is inside polygon, false otherwise
     */
    static func isPointInPolygon(
        point: CLLocationCoordinate2D,
        polygon: [CLLocationCoordinate2D]
    ) -> Bool {
        guard !polygon.isEmpty else { return false }

        // Check if polygon crosses antimeridian
        if crossesAntimeridian(polygon) {
            // Normalize all longitudes to [0, 360] range
            let normalizedPolygon = polygon.map { coord -> CLLocationCoordinate2D in
                return CLLocationCoordinate2D(latitude: coord.latitude, longitude: normalizeLongitude(coord.longitude))
            }
            let normalizedPoint = CLLocationCoordinate2D(latitude: point.latitude, longitude: normalizeLongitude(point.longitude))
            return isPointInPolygonRaw(point: normalizedPoint, polygon: normalizedPolygon)
        }

        return isPointInPolygonRaw(point: point, polygon: polygon)
    }

    /**
     * Raw ray-casting algorithm without antimeridian handling.
     * Internal helper for isPointInPolygon.
     */
    private static func isPointInPolygonRaw(
        point: CLLocationCoordinate2D,
        polygon: [CLLocationCoordinate2D]
    ) -> Bool {
        var intersections = 0
        let x = point.longitude
        let y = point.latitude

        for i in 0..<polygon.count {
            let p1 = polygon[i]
            let p2 = polygon[(i + 1) % polygon.count]

            if (((p1.latitude > y) != (p2.latitude > y)) &&
                (x < (p2.longitude - p1.longitude) * (y - p1.latitude) / (p2.latitude - p1.latitude) + p1.longitude)) {
                intersections += 1
            }
        }

        return intersections % 2 == 1
    }

    /**
     * Distance from a point to a line segment in meters.
     * Uses cosLat-adjusted flat-earth projection for accurate distance calculation.
     * Handles antimeridian-crossing segments by normalizing longitudes.
     *
     * - Parameters:
     *   - p: Point coordinate
     *   - a: Start point coordinate
     *   - b: End point coordinate
     * - Returns: Distance in meters
     */
    static func pointToSegmentDistance(
        p: CLLocationCoordinate2D,
        a: CLLocationCoordinate2D,
        b: CLLocationCoordinate2D
    ) -> Double {
        // Check if segment crosses antimeridian
        let lngDiff = abs(b.longitude - a.longitude)
        if lngDiff > 180.0 {
            // Normalize all longitudes to [0, 360] range
            let normalizedP = CLLocationCoordinate2D(latitude: p.latitude, longitude: normalizeLongitude(p.longitude))
            let normalizedA = CLLocationCoordinate2D(latitude: a.latitude, longitude: normalizeLongitude(a.longitude))
            let normalizedB = CLLocationCoordinate2D(latitude: b.latitude, longitude: normalizeLongitude(b.longitude))
            return pointToSegmentDistanceRaw(p: normalizedP, a: normalizedA, b: normalizedB)
        }

        return pointToSegmentDistanceRaw(p: p, a: a, b: b)
    }

    /**
     * Raw point-to-segment distance calculation without antimeridian handling.
     * Internal helper for pointToSegmentDistance.
     */
    private static func pointToSegmentDistanceRaw(
        p: CLLocationCoordinate2D,
        a: CLLocationCoordinate2D,
        b: CLLocationCoordinate2D
    ) -> Double {
        let ab = haversineDistance(point1: a, point2: b)
        if ab < 0.001 { return haversineDistance(point1: p, point2: a) }

        // Flat-earth projection for segment math (valid for short segments)
        let cosLat = cos((a.latitude + b.latitude) / 2.0 * .pi / 180.0)
        let dx = (b.longitude - a.longitude) * .pi / 180.0 * cosLat * EARTH_RADIUS_METERS
        let dy = (b.latitude - a.latitude) * .pi / 180.0 * EARTH_RADIUS_METERS
        let px = (p.longitude - a.longitude) * .pi / 180.0 * cos((a.latitude + p.latitude) / 2.0 * .pi / 180.0) * EARTH_RADIUS_METERS
        let py = (p.latitude - a.latitude) * .pi / 180.0 * EARTH_RADIUS_METERS

        let abLenSq = dx * dx + dy * dy
        if abLenSq < 0.001 { return haversineDistance(point1: p, point2: a) }

        let t = max(0.0, min(1.0, (px * dx + py * dy) / abLenSq))
        let projLat = a.latitude + t * (b.latitude - a.latitude)
        let projLng = a.longitude + t * (b.longitude - a.longitude)

        return haversineDistance(
            lat1: p.latitude,
            lng1: p.longitude,
            lat2: projLat,
            lng2: projLng
        )
    }
}

File: android/src/main/kotlin/io/polyfence/core/utils/GeoMath.kt
package io.polyfence.core.utils

import kotlin.math.*

/**
 * Centralized geospatial mathematics utility.
 * Single source of truth for geofencing algorithms across the engine.
 * Ensures algorithm parity with iOS implementation.
 *
 * Thread-safe: all functions are pure (no side effects).
 */
internal object GeoMath {
    private const val EARTH_RADIUS_METERS = 6371000.0

    data class LatLng(val latitude: Double, val longitude: Double)

    /**
     * Check if a polygon crosses the antimeridian (±180° longitude).
     * A polygon crosses if any consecutive edge spans >180° longitude difference.
     *
     * @param polygon List of (latitude, longitude) pairs
     * @return true if polygon crosses antimeridian, false otherwise
     */
    private fun crossesAntimeridian(polygon: List<Pair<Double, Double>>): Boolean {
        for (i in polygon.indices) {
            val p1 = polygon[i]
            val p2 = polygon[(i + 1) % polygon.size]
            val lngDiff = Math.abs(p2.second - p1.second)
            if (lngDiff > 180.0) {
                return true
            }
        }
        return false
    }

    /**
     * Normalize longitude from [-180, 180] to [0, 360].
     * Used for antimeridian-crossing polygon handling.
     *
     * @param lng Longitude in [-180, 180]
     * @return Longitude in [0, 360]
     */
    private fun normalizeLongitude(lng: Double): Double {
        return if (lng < 0) lng + 360.0 else lng
    }

    /**
     * Calculate distance between two points using Haversine formula.
     *
     * Reference: Sinnott, R.W. (1984). "Virtues of the Haversine."
     * Sky & Telescope 68(2): 158-159.
     *
     * @param lat1 First point latitude
     * @param lng1 First point longitude
     * @param lat2 Second point latitude
     * @param lng2 Second point longitude
     * @return Distance in meters
     */
    fun haversineDistance(lat1: Double, lng1: Double, lat2: Double, lng2: Double): Double {
        val dLat = Math.toRadians(lat2 - lat1)
        val dLng = Math.toRadians(lng2 - lng1)

        val a = sin(dLat / 2).pow(2) +
                cos(Math.toRadians(lat1)) * cos(Math.toRadians(lat2)) *
                sin(dLng / 2).pow(2)

        val c = 2 * atan2(sqrt(a), sqrt(1 - a))
        return EARTH_RADIUS_METERS * c
    }

    /**
     * Convenience overload: calculate distance between two LatLng points.
     */
    fun haversineDistance(point1: LatLng, point2: LatLng): Double {
        return haversineDistance(point1.latitude, point1.longitude, point2.latitude, point2.longitude)
    }

    /**
     * Point-in-polygon detection using ray casting algorithm.
     * Handles antimeridian-crossing polygons by normalizing longitudes.
     *
     * Reference: Shimrat, M. (1962). "Algorithm 112: Position of point relative to polygon."
     * Communications of the ACM 5(8): 434. DOI: 10.1145/368637.368653
     *
     * @param lat Point latitude
     * @param lng Point longitude
     * @param polygon List of polygon vertices as (lat, lng) pairs
     * @return true if point is inside polygon, false otherwise
     */
    fun isPointInPolygon(lat: Double, lng: Double, polygon: List<Pair<Double, Double>>): Boolean {
        if (polygon.isEmpty()) return false

        // Check if polygon crosses antimeridian
        if (crossesAntimeridian(polygon)) {
            // Normalize all longitudes to [0, 360] range
            val normalizedPolygon = polygon.map { (lat, lng) ->
                Pair(lat, normalizeLongitude(lng))
            }
            val normalizedTestLng = normalizeLongitude(lng)
            return isPointInPolygonRaw(lat, normalizedTestLng, normalizedPolygon)
        }

        return isPointInPolygonRaw(lat, lng, polygon)
    }

    /**
     * Raw ray-casting algorithm without antimeridian handling.
     * Internal helper for isPointInPolygon.
     */
    private fun isPointInPolygonRaw(lat: Double, lng: Double, polygon: List<Pair<Double, Double>>): Boolean {
        var intersections = 0
        val x = lng
        val y = lat

        for (i in polygon.indices) {
            val p1 = polygon[i]
            val p2 = polygon[(i + 1) % polygon.size]

            if (((p1.first > y) != (p2.first > y)) &&
                (x < (p2.second - p1.second) * (y - p1.first) / (p2.first - p1.first) + p1.second)) {
                intersections++
            }
        }

        return intersections % 2 == 1
    }

    /**
     * Convenience overload: point-in-polygon with LatLng point and polygon.
     */
    fun isPointInPolygon(point: LatLng, polygon: List<LatLng>): Boolean {
        val pairPolygon = polygon.map { Pair(it.latitude, it.longitude) }
        return isPointInPolygon(point.latitude, point.longitude, pairPolygon)
    }

    /**
     * Distance from a point to a line segment in meters.
     * Uses cosLat-adjusted flat-earth projection for accurate distance calculation.
     * Handles antimeridian-crossing segments by normalizing longitudes.
     * Distance calculation via vector projection (classical geometric method).
     *
     * @param pLat Point latitude
     * @param pLng Point longitude
     * @param aLat Start point latitude
     * @param aLng Start point longitude
     * @param bLat End point latitude
     * @param bLng End point longitude
     * @return Distance in meters
     */
    fun pointToSegmentDistance(
        pLat: Double,
        pLng: Double,
        aLat: Double,
        aLng: Double,
        bLat: Double,
        bLng: Double
    ): Double {
        // Check if segment crosses antimeridian
        val lngDiff = Math.abs(bLng - aLng)
        if (lngDiff > 180.0) {
            // Normalize all longitudes to [0, 360] range
            val normalizedPLng = normalizeLongitude(pLng)
            val normalizedALng = normalizeLongitude(aLng)
            val normalizedBLng = normalizeLongitude(bLng)
            return pointToSegmentDistanceRaw(pLat, normalizedPLng, aLat, normalizedALng, bLat, normalizedBLng)
        }

        return pointToSegmentDistanceRaw(pLat, pLng, aLat, aLng, bLat, bLng)
    }

    /**
     * Raw point-to-segment distance calculation without antimeridian handling.
     * Internal helper for pointToSegmentDistance.
     */
    private fun pointToSegmentDistanceRaw(
        pLat: Double,
        pLng: Double,
        aLat: Double,
        aLng: Double,
        bLat: Double,
        bLng: Double
    ): Double {
        val ab = haversineDistance(aLat, aLng, bLat, bLng)
        if (ab < 0.001) return haversineDistance(pLat, pLng, aLat, aLng)

        // Flat-earth projection for segment math (valid for short segments)
        val cosLat = cos(Math.toRadians((aLat + bLat) / 2))
        val dx = Math.toRadians(bLng - aLng) * cosLat * EARTH_RADIUS_METERS
        val dy = Math.toRadians(bLat - aLat) * EARTH_RADIUS_METERS
        val px = Math.toRadians(pLng - aLng) * cos(Math.toRadians((aLat + pLat) / 2)) * EARTH_RADIUS_METERS
        val py = Math.toRadians(pLat - aLat) * EARTH_RADIUS_METERS

        val abLenSq = dx * dx + dy * dy
        if (abLenSq < 0.001) return haversineDistance(pLat, pLng, aLat, aLng)

        val t = ((px * dx + py * dy) / abLenSq).coerceIn(0.0, 1.0)

        val projLat = aLat + t * (bLat - aLat)
        val projLng = aLng + t * (bLng - aLng)

        return haversineDistance(pLat, pLng, projLat, projLng)
    }

    /**
     * Convenience overload: point-to-segment with LatLng points.
     */
    fun pointToSegmentDistance(p: LatLng, a: LatLng, b: LatLng): Double {
        return pointToSegmentDistance(p.latitude, p.longitude, a.latitude, a.longitude, b.latitude, b.longitude)
    }

    /**
     * Minimum distance from a point to a polygon boundary in meters.
     * Iterates over all edges of the polygon and returns the smallest point-to-segment distance.
     *
     * @param pLat Point latitude
     * @param pLng Point longitude
     * @param polygon List of (latitude, longitude) pairs forming the polygon
     * @return Minimum distance in meters from the point to any polygon edge
     */
    fun pointToPolygonDistance(pLat: Double, pLng: Double, polygon: List<Pair<Double, Double>>): Double {
        if (polygon.size < 2) return Double.MAX_VALUE
        var minDist = Double.MAX_VALUE
        for (i in polygon.indices) {
            val j = (i + 1) % polygon.size
            val dist = pointToSegmentDistance(
                pLat, pLng,
                polygon[i].first, polygon[i].second,
                polygon[j].first, polygon[j].second
            )
            if (dist < minDist) minDist = dist
        }
        return minDist
    }
}
